File: webapps/api/backend/views/edgeclients.py
# Copyright (C) 2016 iNuron NV
#
# This file is part of Open vStorage Open Source Edition (OSE),
# as available from
#
#      http://www.openvstorage.org and
#      http://www.openvstorage.com.
#
# This file is free software; you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
# as published by the Free Software Foundation, in version 3 as it comes
# in the LICENSE.txt file of the Open vStorage OSE distribution.
#
# Open vStorage is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY of any kind.

"""
Module for domains
"""

from backend.decorators import load, log, required_roles, return_list
from backend.exceptions import HttpNotImplementedException, HttpNotAcceptableException
from ovs.dal.hybrids.edgeclient import EdgeClient
from ovs.dal.lists.storagedriverlist import StorageDriverList
from ovs.dal.lists.vdisklist import VDiskList
from rest_framework import viewsets
from rest_framework.permissions import IsAuthenticated


class EdgeClientViewSet(viewsets.ViewSet):
    """
    Information about Domains
    """
    permission_classes = (IsAuthenticated,)
    prefix = r'edgeclients'
    base_name = 'edgeclients'

    @log()
    @required_roles(['read'])
    @return_list(EdgeClient)
    @load()
    def list(self, volume_id=None):
        """
        Lists all available EdgeClients
        """
        clients = []
        if volume_id is None:
            storagedrivers = StorageDriverList.get_storagedrivers()
        else:
            vdisk = VDiskList.get_vdisk_by_volume_id(volume_id)
            if vdisk is None:
                raise HttpNotAcceptableException(error_description='The given volume_id does not link to a vDisk',
                                                 error='object_not_found')
            storagedrivers = vdisk.vpool.storagedrivers
        for sd in storagedrivers:
            for item in sd.vpool.storagedriver_client.list_client_connections(str(sd.storagedriver_id)):
                if volume_id is not None and item.object_id != volume_id:
                    continue
                client = EdgeClient(data={'object_id': item.object_id,
                                          'ip': item.ip,
                                          'port': item.port},
                                    volatile=True)
                clients.append(client)
        clients.sort(key=lambda e: ('ip', 'port'))
        return clients

    @log()
    @required_roles(['read'])
    @load()
    def retrieve(self):
        """
        Load information about a given Domain
        """
        raise HttpNotImplementedException(error_description='EdgeClients cannot be requested individually',
                                          error='unavailable')

File: webapps/frontend/app/viewmodels/site/edgeclients.js
// Copyright (C) 2016 iNuron NV
//
// This file is part of Open vStorage Open Source Edition (OSE),
// as available from
//
//      http://www.openvstorage.org and
//      http://www.openvstorage.com.
//
// This file is free software; you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
// as published by the Free Software Foundation, in version 3 as it comes
// in the LICENSE.txt file of the Open vStorage OSE distribution.
//
// Open vStorage is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY of any kind.
/*global define, window */
define([
    'jquery', 'plugins/dialog', 'knockout',
    'ovs/shared', 'ovs/generic', 'ovs/refresher', 'ovs/api',
    '../containers/vdisk', '../containers/edgeclient'
], function($, dialog, ko, shared, generic, Refresher, api, VDisk, EdgeClient) {
    "use strict";
    return function() {
        var self = this;

        // Variables
        self.shared            = shared;
        self.guard             = { authenticated: true };
        self.vDiskCache        = {};
        self.clientLookup      = {};
        self.widgets           = [];
        self.edgeClientHeaders = [
            { key: 'vdisk', value: $.t('ovs:generic.vdisk'), width: undefined },
            { key: 'ip',    value: $.t('ovs:generic.ip'),    width: 200       },
            { key: 'port',  value: $.t('ovs:generic.port'),  width: 100       }
        ];

        // Handles
        self.edgeClientHandle = {};

        // Observables
        self.edgeClients = ko.observableArray([]);

        // Functions
        self.loadEdgeClients = function(options) {
            return $.Deferred(function(deferred) {
                if (generic.xhrCompleted(self.edgeClientHandle[options.page])) {
                    options.contents = 'object_id';
                    self.edgeClientHandle[options.page] = api.get('edgeclients', { queryparams: options })
                        .then(function(data) {
                            deferred.resolve({
                                data: data,
                                loader: function(guid) {
                                    return new EdgeClient(guid);
                                },
                                dependencyLoader: function(item) {
                                    var id = item.objectId();
                                    if (!self.vDiskCache.hasOwnProperty(id)) {
                                        self.vDiskCache[id] = undefined;
                                    } else {
                                        item.vDisk(self.vDiskCache[id]);
                                    }
                                }
                            });
                        })
                        .done(function() {
                            var volumes = [], query;
                            $.each(self.vDiskCache, function(objectId, vdisk) {
                                if (vdisk === undefined) {
                                    volumes.push(objectId);
                                }
                            });
                            query = {
                                type: 'AND',
                                items: [['volume_id', 'IN', volumes]]
                            };
                            return api.get('vdisks', { queryparams: { query: JSON.stringify(query), contents: 'name' }})
                                .done(function(data) {
                                    $.each(data.data, function(index, entry) {
                                        if (self.vDiskCache[entry.volume_id] === undefined) {
                                            self.vDiskCache[entry.volume_id] = new VDisk(entry.guid);
                                        }
                                        self.vDiskCache[entry.volume_id].fillData(entry);
                                    });
                                    $.each(self.edgeClients(), function(index, client) {
                                        if (client.vDisk() === undefined) {
                                            client.vDisk(self.vDiskCache[client.objectId()]);
                                        }
                                    })
                                });
                        })
                        .fail(function() { deferred.reject(); });
                } else {
                    deferred.resolve();
                }
            }).promise();
        };

        // Durandal
        self.deactivate = function() {
            $.each(self.widgets, function(i, item) {
                item.deactivate();
            });
        };
    };
});
